[DOCS] Expands ELSER tutorial with optimization info

szabosteve opened the pull request:
## Overview

This PR:
* adds a new section to the ELSER tutorial about how to save disk space by excluding `_source` from the destination index,
* adds a search performance optimization section to the `text_expansion` query documentation,
* links the tutorial and the new optimization section of `text_expansion`,
* clarifies which field contains the generated tokens in the mappings of the dest index.

### Preview

* [ELSER tutorial](https://elasticsearch_97392.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/semantic-search-elser.html)
* [text_expansion - performance optimization](https://elasticsearch_97392.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/query-dsl-text-expansion-query.html#optimizing-text-expansion)

szabosteve committed:
[DOCS] Adds section about saving space with excluding tokens from source.

szabosteve committed:
[DOCS] Adds search performance optimization section to text expansion docs.

szabosteve committed:
[DOCS] Adds link to the ELSER tutorial that points to the optimization section of the text_expansion query.

szabosteve committed:
[DOCS] Fine-tunes text.

elasticsearchmachine commented:
Pinging @elastic/ml-core (Team:ML)

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

davidkyle commented on docs/reference/query-dsl/text-expansion-query.asciidoc:
```diff
@@ -76,4 +76,31 @@ GET my-index/_search
+[[optimizing-text-expansion]]
+=== Optimizing the search performance of the text_expansion query
+
+https://www.elastic.co/blog/faster-retrieval-of-top-hits-in-elasticsearch-with-block-max-wand[Max WAND] 
```
```suggestion
https://www.elastic.co/blog/faster-retrieval-of-top-hits-in-elasticsearch-with-block-max-wand[Max WAND] is an
```

davidkyle commented on docs/reference/query-dsl/text-expansion-query.asciidoc:
```diff
@@ -76,4 +76,31 @@ GET my-index/_search
+=== Optimizing the search performance of the text_expansion query
+
+https://www.elastic.co/blog/faster-retrieval-of-top-hits-in-elasticsearch-with-block-max-wand[Max WAND] 
+optimization technique used by {es} to skip over documents that don't score 
```
```suggestion
optimization technique used by {es} to skip documents that cannot score 
```

davidkyle commented on docs/reference/query-dsl/text-expansion-query.asciidoc:
```diff
@@ -76,4 +76,31 @@ GET my-index/_search
+deciding which setting to use.
+
+NOTE: The `track_total_hits` option applies to all queries in the search request 
+and may be optimal for some queries but not for others. Take into account of the 
```
```suggestion
and may be optimal for some queries but not for others. Take into account the 
```

davidkyle commented on docs/reference/query-dsl/text-expansion-query.asciidoc:
```diff
@@ -76,4 +76,31 @@ GET my-index/_search
+latency for `text_expansion`. For datasets of a significant size, disabling Max 
+WAND leads to lower query latencies.
+
+The behavior of Max WAND can be controlled using the 
```
```suggestion
Max WAND is controlled by the
```

davidkyle commented on docs/reference/query-dsl/text-expansion-query.asciidoc:
```diff
@@ -76,4 +76,31 @@ GET my-index/_search
+https://www.elastic.co/blog/faster-retrieval-of-top-hits-in-elasticsearch-with-block-max-wand[Max WAND] 
+optimization technique used by {es} to skip over documents that don't score 
+competitively against the current best matching documents. However, the weighted 
+terms generated by the ELSER model don't work well with the Max WAND 
```
The rest of the doc uses the phrase `token-weight pairs` rather than `weighted terms`, to be consistent `tokens` should be preferred to `terms`.

This could be rewritten as "However, the tokens generated by the ELSER...."

davidkyle commented on docs/reference/search/search-your-data/semantic-search-elser.asciidoc:
```diff
@@ -49,31 +49,78 @@ to index the ELSER output.
+
+[discrete]
+[[save-space]]
+==== Saving disk space by configuring the destination index
```
```suggestion
==== Saving disk space by excluding the ELSER tokens from document source
```

davidkyle commented on docs/reference/search/search-your-data/semantic-search-elser.asciidoc:
```diff
@@ -49,31 +49,78 @@ to index the ELSER output.
+[[save-space]]
+==== Saving disk space by configuring the destination index
+
+The weighted terms generated by ELSER must be indexed for use in the 
```
Same point about using `tokens` instead of `terms`

davidkyle commented on docs/reference/search/search-your-data/semantic-search-elser.asciidoc:
```diff
@@ -49,31 +49,78 @@ to index the ELSER output.
+consider this trade-off and make sure that excluding the ELSER terms from the 
+source aligns with your specific requirements and use case.
+
+The mapping that excludes `_source` from the destination index can be created by 
```
```suggestion
The mapping that excludes `ml.tokens` from the  `_source` field can be created by 
```

davidkyle commented on docs/reference/search/search-your-data/semantic-search-elser.asciidoc:
```diff
@@ -49,31 +49,78 @@ to index the ELSER output.
+
+
+[discrete]
+[[save-space]]
```
Can this section be moved to the bottom of the file? I feel like it is a digression from the flow of the tutorial and is something to consider once the user has got started with ELSER

szabosteve committed:
Apply suggestions from code review

Co-authored-by: David Kyle <david.kyle@elastic.co>

szabosteve committed:
[DOCS] Addresses feedback.

szabosteve commented:
@davidkyle I addressed your feedback, thanks for those. I added a section called `Optimizing performance` to the ELSER tutorial which includes `Saving disk space by excluding the ELSER tokens from document source`. Please let me know what you think. If we have more optimization content, I'll put it on a separate page.

szabosteve committed:
Merge branch 'main' into elser-space

jimczi reviewed (approved):
lgtm

davidkyle reviewed (approved):
LGTM

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.9  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.8  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 97392`

szabosteve commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.9|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/97784">](https://github.com/elastic/elasticsearch/pull/97784)|
|✅|8.8|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/97785">](https://github.com/elastic/elasticsearch/pull/97785)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.8 -->
